### DISCUSSION TITLE
How to get number of connections?

### DISCUSSION THREAD

**Discussion by khvzak** [ntex-rs/ntex#181] [Q&A] [answered]
Hello
Is it possible to calculate _total_ number of connections received, and number of active (in-flight) connections?  (per worker or global) 
This is all useful for application metrics to analyze performance, find issues, and so on.

In actix-web it was possible using the `on_connect` callback and storing a counter (with destructor) in extensions container.

**Comment by fakeshadow** [comment]
It would be better to use external tooling for runtime profiling.
Since you brought up actix-web just look at it's recent issues where a simple example can trigger connection hang. You would have no (easy) way to profiling or even noticing it within the library itself.

**Comment by fafhrd91** [comment, marked as answer]
@khvzak you can implement middleware service `ServiceFactory<Io<F>>` and wrap http service in it, then you’ll be able to access Io at the begging and after http service completes handling

regarding actix-web and errors, that shouldn’t be too big of surprise. it’s http1 dispatcher is mostly one that i wrote several years ago and it was shit, nothing changed recently. i rewrote dispatcher in ntex 3 times until settle with current design.

**Comment by khvzak** [comment]
Thanks Nikolay!

I'll try to add a middleware to wrap http service in.

The ntex design (incl. middlewares, http body, etc) looks better than in actix (and more easy to use). I'm trying to migrate from actix (as an experiment) in my company's project and see how it will go.

UPD: 
It works, code snippet:
```rust
let service =
    HttpService::build()
        .finish(fn_service(|_req| async move {
            Ok::<_, std::io::Error>(Response::Ok().body("Hello world!"))
        }));

apply_fn_factory(service, |io: Io, handler| {
    let peer_addr = io.query::<PeerAddr>().get().unwrap();
    println!("new connection from {peer_addr:?}");
    handler.call(io)
})
```
